src/core/tsi/ssl_transport_security.cc

```diff
@@ -490,21 +490,35 @@ static tsi_result peer_from_x509(X509* cert, int include_certificate_type,
   return result;
 }
 
-/* Logs the SSL error stack. */
-static void log_ssl_error_stack(void) {
-  unsigned long err;
-  while ((err = ERR_get_error()) != 0) {
-    char details[256];
-    ERR_error_string_n(static_cast<uint32_t>(err), details, sizeof(details));
-    gpr_log(GPR_ERROR, "%s", details);
+/* Retrieves the SSL error stack as a string. */
+static std::string ssl_error_stack_to_string(void) {
+  BIO* bio = BIO_new(BIO_s_mem());
+  ERR_print_errors(bio);
```

Comment:
In order to keep a consistent logging story for gRPC, we probably want to use `gpr_log()` here.
Can we keep the original loop for clearing error queue?

Reply:
I found the current setup sort of unfortunate for two reasons:

* ERR_error_string (_n) truncates the error messages.

* The root cause is printed in a separate log statement from the "corruption detected" message.  As an example, it doesn't really make any sense that failure adding a directory full of X.509 trust anchors could lead to "corruption":
```
W20210729 23:07:00.135056 23154688 ssl_transport_security.cc:510] Corruption detected.
W20210729 23:07:00.135092 23154688 ssl_transport_security.cc:486] error:0B064071:x509 certificate routines:add_cert_dir:invalid directory
```

IMO this issue would've been much easier to debug if these had been compacted into a single statement.

Reply:
Is it possible to concatenate all the error messages into one line using a for-loop, and then log them all in once by using `gpr_log()`? Or alternatively, is it possible to read all the characters from bio, and we then log it through `gpr_log()`?
I could see there is a need for changing the way we log. But is there a way we can avoid using OpenSSL's built-in logging methods?

Reply:
[ERR_print_errors](https://commondatastorage.googleapis.com/chromium-boringssl-docs/bio.h.html#ERR_print_errors) prints to the BIO, not to stderr or stdout.  In this case, we created the BIO with [BIO_s_mem](https://commondatastorage.googleapis.com/chromium-boringssl-docs/bio.h.html#BIO_s_mem), so it refers to an in-memory buffer.  Which is then copied to a `std::string` for subsequent logging with `gpr_log()`.

Reply:
Ah I see, that would be perfect. Thanks for the explanation!